### DISCUSSION TITLE
How to include mongoose web server and freertos

### DISCUSSION THREAD

**Discussion by ghost** [earlephilhower/arduino-pico#1958] [General]
I'm trying to include mongoose web server and freertos but there are build issues with LWIP opts.
Could you help me to solve this issue sharing a platformio.ini example?

**Comment by earlephilhower** [comment]
Moved to discussion since this seems like a "how to XXX" query.

LWIP options are fixed by the core itself and can't be changed or redefined by an app.  LWIP comes as part of the Pico SDK and is used by the PicoW driver to integrate networking at that level.

If that's not what you're trying to do then posting some code and error messages would be helpful.

**Reply by maxgerhardt** [reply]
While there is an example <https://mongoose.ws/documentation/tutorials/rp2040/pico-w/> and <https://github.com/mongoose-examples/raspberry-pi-pico-w/tree/main> against Pico-SDK, it does modify the lwipopts.h. Specifically, Mongoose requires `LWIP_SOCKET` = 1 because it wants to have the regular socket API and defines instead of talking to the direct lwIP API.

**Reply by maxgerhardt** [reply]
This thing has one "gotcha" after the other.
First of all, Mongoose requires LWIP_SOCKET = 1. Okay. But, when you activate that, it immediately errors out with `NO_SYS must be 0`. So I put it to `0`. But then it needs to know about the underlying "OS". Okay, FreeRTOS it is. I modify the CMake to link in `pico_lwip_freertos` instead of `pico_lwip_nosys`. But that blows it up with `#error PICO_CYW43_ARCH_THREADSAFE_BACKGROUND requires lwIP NO_SYS=1`. I have a feeling the integratino of the CYW43 driver with LWIP very much relies on NO_SYS=1, while mongoose through its socket API requirements needs the opposite.

(Oh and it also needs `LWIP_TIMEVAL_PRIVATE=0` otherwise some timeval structure conflicts with the toolchain.)

**Reply by earlephilhower** [reply]
Yeah, the `LWIP_SOCKET` API is a non-starter here.  You will need to use bare metal SDK to build and run that program, it's not related to the Arduino ecosystem in any way...

**Reply by maxgerhardt** [reply]
I've created an unholy mess by modifying the CMakeLists.txt to satisfy `NO_SYS=0`, `LWIP_SOCKET=1` and sprinkling in some compilation fixes, but the resulting binary just crashes and loses the USB serial port, before even entering `setup()`. (for anyone interested: https://github.com/maxgerhardt/pico-lwip-messing-around).

I think the approach must be rethought if Mongoose is to be usable in this core, e.g., Mongoose should be tought how to use the native lwIP API that is available in the configuration this core uses, or, maybe even the already abstracted `WiFiServer`, `WiFiClient` libraries / classes..

**Reply by ghost** [reply]
Thanks for your help. Supporting Mongoose would be a great achievement. Unfortunately I'm not so expert on MCUs to help this project :disappointed:.

**Reply by maxgerhardt** [reply]
If it's just about having "a" webserver, you can use the current [WebServer](https://github.com/earlephilhower/arduino-pico/tree/master/libraries/WebServer/examples) library just fine. It probably has all you need already.

**Reply by ghost** [reply]
I need to stream data as fast as possible using websockets. I tried [arduinoWebSockets](https://github.com/Links2004/arduinoWebSockets) project but it stops working properly after a non-reproducible time. I tried to enable debugging of the project but when the client try to connect it receive a bad HTTP error message.

This is the reason why I was trying to integrate Mongoose.

**Reply by maxgerhardt** [reply]
Then I'd propose you try the Mongoose example for the Pico-SDK (as linked above with `mongoose-examples/raspberry-pi-pico-w`) with some fake data generation to see if you run in the same issue.

**Reply by ghost** [reply]
Thanks for your suggestion, I'll try it asap and update you.

Independenty from the results I would like to keep both ways (CMakeLists.txt + pico-sdk + freertos + mongoose) and (platformio.ini + arduino-pico + arduinoWebSockets). Have you ever tried? Have you a sample project? I like the PlatformIO project structure (src + lib + test + include + ...) because it is very clear.

Thanks in advance :smile:
